cmd/main.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"flag"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/spf13/pflag"
+	"k8s.io/klog/v2"
+
+	"kubertas/dbreader"
+	"kubertas/knode"
+	"kubertas/utils"
+)
+
+var (
+	wg          sync.WaitGroup
+	cfgFilePath string
+)
+
+func main() {
+	klogFlags := flag.NewFlagSet("klog", flag.ExitOnError)
+	klog.InitFlags(klogFlags)
+	klogFlags.Set("logtostderr", "true")
+
+	pflag.CommandLine.AddGoFlagSet(klogFlags)
+	pflag.StringVarP(&cfgFilePath, "config", "c", "config.json", "Path to KubeRTAS config file")
+	pflag.Parse()
+
+	klog.V(0).Infof("setting up KubeRTAS, reading config from %s", cfgFilePath)
+	serviceConfig, err := utils.ReadConfig(cfgFilePath)
+	if err != nil {
+		klog.Errorf("error while reading config. Error: %v", err)
+		return
+	}
+
+	// Check if the servicelog.db file is present in the path set in config file.
+	if _, err := os.Stat(serviceConfig.ServicelogPath); err != nil {
+		klog.Errorf("Error finding the servicelog.db file. %v", err)
+		return
+	}
+
+	// Create an instance of NewReader to retieve entries from the servicelog.db file.
+	svLogReader := dbreader.NewReader(serviceConfig.ServicelogPath, serviceConfig.Severity)
+
+	// Set up notifier to post events to the Kube API server.
+	notifier := knode.NewK8sNotifier(serviceConfig.InCluster)
+	err = notifier.InitializeK8sNotifier()
+	if err != nil {
+		klog.Errorf("cannot initialize the k8s apiserver notifier. %v", err)
+		return
+	}
+
+	// Set up a ticker to periodically check the servicelog.db file for new entries.
+	ticker := time.NewTicker(serviceConfig.PollInterval * time.Second)
+	done := make(chan struct{}, 1)
+	initRun := make(chan struct{}, 1)
+	wg.Add(2)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				ticker.Stop()
+				klog.V(1).Info("Ticker stopped.")
+				wg.Done()
+			case <-ticker.C:
+			case <-initRun:
+			}
+			if err = svLogReader.ReadServicelogDB(notifier); err != nil {
+				klog.Errorf("error while reading the logs from the servicelog.db file. %v", err)
+				return
+			}
+		}
+	}()
+
+	// Capture any OS signals to perform graceful exit.
+	sigchan := make(chan os.Signal, 1)
+	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		sig := <-sigchan
+		done <- struct{}{}
+		klog.Infof("Captured %s, Shutting down KubeRTAS", sig.String())
+		wg.Done()
+		os.Exit(0)
```

Comment:
wont it automatically exit the program after wg.Done(), is it used to be on safer side?

Reply:
@Karthik-K-N, It will automatically exit. I've kept it nevertheless.